pandas/core/case_when.py

```diff
@@ -0,0 +1,151 @@
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when(df.a.gt(0), df.a,   # condition, replacement
+    ...              df.b.gt(0), df.b,
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
+    len_args = len(args)
+    if not len_args:
+        raise ValueError(
+            "Kindly provide at least one boolean condition, "
+            "with a corresponding replacement."
+        )
+    if len_args % 2:
+        raise ValueError(
+            "The number of boolean conditions should be equal "
+            "to the number of replacements. "
+            "However, the total number of conditions and replacements "
+            f"is {len(args)}, which is an odd number."
+        )
+
+    conditions = args[-2::-2]
+    replacements = args[-1::-2]
+    common_dtype = [infer_dtype_from(replacement)[0] for replacement in replacements]
+    if default is not None:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtype.append(arg_dtype)
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            construct_1d_arraylike_from_scalar(
+                value=replacement, length=len(condition), dtype=common_dtype
+            )
+            if is_scalar(replacement)
+            else replacement.astype(common_dtype, copy=False)
+            if isinstance(replacement, ABCSeries)
+            else pd_array(replacement, dtype=common_dtype, copy=False)
+            for condition, replacement in zip(conditions, replacements)
+        ]
+        if (default is not None) and isinstance(default, ABCSeries):
+            default = default.astype(common_dtype, copy=False)
+    else:
+        common_dtype = common_dtype[0]
+    # TODO: possibly extend this to a DataFrame?
+    if not isinstance(default, ABCSeries):
+        default_index = len(args[0])
+        for condition in conditions[::-1]:
+            if isinstance(condition, ABCSeries):
+                default_index = condition.index
+                break
+        default = Series(default, index=default_index, dtype=common_dtype)
+    counter = len_args // 2 - 1
+    counter = range(counter, -1, -1)
+    for position, condition, replacement in zip(counter, conditions, replacements):
+        try:
+            default = default.mask(
+                condition, other=replacement, axis=0, inplace=False, level=level
+            )
+        except Exception as error:
```

Comment:
I would remove this try except and have mask raise it's error normally

Reply:
The idea here is to keep track of which `condition, replacement` failed. More like `condition1` failed, this is why it failed. Devolving to mask error directly and you lose the error tracking. I assume the error tracking would be useful to the user.